specreduce/utils/utils.py

```diff
@@ -0,0 +1,170 @@
+import copy
+
+import numpy as np
+
+from specreduce.tracing import Trace, FlatTrace
+from specreduce.extract import _ap_weight_image
+
+
+def measure_cross_dispersion_profile(image, trace=None, crossdisp_axis=0,
+                                     width=None, pixel=None, pixel_range=None,
+                                     statistic='median'):
+    """
+    Return a 1D (quantity) array of the cross-dispersion profile measured at a
+    specified pixel value ('wavelength', but potentially before calibration),
+    or the average profile across several pixel values (or range of pixel values)
+    along the dispersion axis.
+
+    If a single number is specified for `pixel`, then the profile at that pixel
+    (i.e wavelength) will be returned. If several pixels are specified in a list or
+    array, then they will be averaged (median or mean, set by `statistic` which
+    defaults to median). Alternatively, `pixel_range` can be specified as a tuple
+    of integers specifying the minimum and maximum pixel range to average the
+    profile. `pixel` and `pixel_range` cannot be set simultaneously, and the default
+    is `pixel_range`=(min_image, max_image) to return an average profile across the
+    entire wavelength range of the image.
+
+    The window in the cross dispersion axis for measuring the profile at
+    the pixel(s) specified is determined by `width` and `trace`. If a trace is
+    provided (either as a Trace object, or as a number specifying the location
+    on the cross-dispersion axis of a FlatTrace object) that will determine the
+    center of the profile on the cross-dispersion axis. Otherwise, if `trace`
+    is None, the center of the image will be the center of the returned profile
+    (i.e., the center row assuming a horizontal trace).
+
+    If `width` is none, a window size of half the image in the cross-dispersion
+    axis will be used to measure the cross dispersion profile. Otherwise, an
+    integer value can be set for `width` which will determine the size of the
+    window around the trace used to measure the profile (this window moves with
+    the trace if trace is not flat).
+
+    Parameters
+    ----------
+    image : `~astropy.nddata.NDData`-like or array-like, required
```

Comment:
Trace and Background accept a Spectrum1D object too. Can this work in the same way? If I try feeding a Spectrum1D object, I get this traceback:

MaskError                                 Traceback (most recent call last)
Cell In[19], line 1
----> 1 profile = cdprofile(objectspec, trace)

File [~/opt/miniconda3/envs/prsr214/lib/python3.11/site-packages/specreduce/utils/utils.py:157](http://localhost:8888/~/opt/miniconda3/envs/prsr214/lib/python3.11/site-packages/specreduce/utils/utils.py#line=156), in measure_cross_dispersion_profile(image, trace, crossdisp_axis, width, pixel, pixel_range, statistic)
    154 # combine these masks to isolate the rows and cols used to measure profile
    155 combined_mask = np.logical_or(pixel_mask, wimg)
--> 157 masked_arr = np.ma.MaskedArray(image, combined_mask)
    159 # and measure the cross dispersion profile. if multiple pixels[/wavelengths](http://localhost:8888/wavelengths),
    160 # this will be an average. we already transposed data based on disp_axis so
    161 # axis is always 1 for this calculation
    162 if statistic == 'mean':

File [~/opt/miniconda3/envs/prsr214/lib/python3.11/site-packages/numpy/ma/core.py:2911](http://localhost:8888/~/opt/miniconda3/envs/prsr214/lib/python3.11/site-packages/numpy/ma/core.py#line=2910), in MaskedArray.__new__(cls, data, mask, dtype, copy, subok, ndmin, fill_value, keep_mask, hard_mask, shrink, order)
   2908     else:
   2909         msg = "Mask and data not compatible: data size is %i, " + \
   2910               "mask size is %i."
-> 2911         raise MaskError(msg % (nd, nm))
   2912     copy = True
   2913 # Set the mask to the new value

MaskError: Mask and data not compatible: data size is 1, mask size is 2000.

Reply:
Yeah I am working on this - I need to move what is in _parse_image somewhere else so this can use it